lib/compilers/vb.js

```diff
@@ -0,0 +1,159 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import path from 'path';
+import fs from 'fs-extra';
+
+import { BaseCompiler } from '../base-compiler';
+
+export class VBCompiler extends BaseCompiler {
+
+    static get key() { return 'vb'; }
+
+    get rID() { return this.compilerProps('compiler.vb.runtimeId'); }
+    get targetFramework() { return this.compilerProps('compiler.vb.targetFramework'); }
+    get buildConfig() { return this.compilerProps('compiler.vb.buildConfig'); }
+
+    get compilerOptions() {
+        return ['publish', '-c', this.buildConfig, '--self-contained', '--runtime', this.rID];
+    }
+
+    async runCompiler(compiler, options, inputFileName, execOptions) {
+        if (!execOptions) {
+            execOptions = this.getDefaultExecOptions();
+        }
+
+        const programDir = path.dirname(inputFileName);
+        const clrBuildDir = this.compilerProps('compiler.vb.clrDir');
+
+        const projectFilePath = path.join(programDir, 'CompilerExplorer.vbproj');
+        const coreClrJitPath = path.join(clrBuildDir, this.compilerProps('compiler.vb.jit'));
+        const crossgen2Path = path.join(clrBuildDir, 'crossgen2', this.compilerProps('compiler.vb.crossgen2'));
+
+        const programPublishPath = path.join(
+            programDir,
+            'bin',
+            this.buildConfig,
+            this.targetFramework,
+            this.rID,
+            'publish',
+        );
+
+        const programDllPath = path.join(programPublishPath, 'CompilerExplorer.dll');
+
+        const projectFileContent =
+            `<Project Sdk="Microsoft.NET.Sdk">
+            <PropertyGroup>
+                <TargetFramework>${this.targetFramework}</TargetFramework>
+                <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+                <Nullable>enable</Nullable>
+                <AssemblyName>CompilerExplorer</AssemblyName>
+            </PropertyGroup>
+         </Project>
+        `;
+
+        execOptions.customCwd = programDir;
+        await fs.writeFile(projectFilePath, projectFileContent);
+        options = this.compilerOptions;
+
+        const compilerResult = await super.runCompiler(
+            compiler,
+            options,
+            inputFileName,
+            execOptions,
+        );
+
+        if (compilerResult.code !== 0)
+            return compilerResult;
+
+        const crossgen2Result = await this.runCrossgen2(
+            execOptions,
+            crossgen2Path,
+            coreClrJitPath,
+            programPublishPath,
+            programDllPath,
+            this.getOutputFilename(programDir, ''),
+        );
+
+        if (crossgen2Result.code !== 0)
+            return crossgen2Result;
+
+        return compilerResult;
+    }
+
+    optionsForFilter() {
+        return this.compilerOptions;
+    }
+
+    getOutputFilename(dirPath) {
+        return path.join(dirPath, `output.s`);
+    }
+
+    cleanAsm(asm) {
```

Comment:
this seems duplicated, can you make this a shared base class function or a shared util function of some kind

Reply:
Done. I've moved all duplicated code to `dotnet.js`.

Reply:
Excellent